src/models/onnxruntime_api.h

```diff
@@ -78,9 +85,72 @@ namespace Ort {
+  // for OrtGetApiBase, and the ORT versions don't match, you will get a cryptic runtime error like:
+  //   java.lang.UnsatisfiedLinkError: dlopen failed: cannot locate symbol "OrtGetApiBase" referenced by
+  //                                   "/data/app/<appname>/base.apk!/lib/x86_64/libonnxruntime-genai.so"
+  //
+  // In order to be flexible we need to add complexity here by:
+  //   - don't call OrtGetApiBase directly so libonnxruntime-genai.so does not have a hard dependency
+  //     on libonnxruntime.so for the symbol.
+  //   - use dlopen/dysm to manually load the onnxruntime library and get the OrtGetApiBase function pointer.
+  //     - requires the Android app has referenced both the GenAI and ORT Android packages so the library is available.
+  //   - iterate between the current ORT version we're aware of, and a minimum required version, so that we work with
+  //     any libonnxruntime.so that supports one of those versions.
+  //
+
+  const std::string path = "libonnxruntime.so";  // "libonnxruntime4j_jni.so" is also an option if we have issues
+  __android_log_print(ANDROID_LOG_INFO, "GenAI", "Attempting to dlopen %s native library", path.c_str());
+
+  using OrtApiBaseFn = const OrtApiBase* (*)(void);
+  OrtApiBaseFn ort_api_base_fn = nullptr;
+
+  void* ort_lib_handle = dlopen(path.c_str(), RTLD_LOCAL);
+  if (ort_lib_handle == nullptr) {
+    __android_log_assert("ort_lib_handle != nullptr", "GenAI", "Failed to load %s", path.c_str());
+  }
+
+  ort_api_base_fn = (OrtApiBaseFn)dlsym(ort_lib_handle, "OrtGetApiBase");
+  if (ort_api_base_fn == nullptr) {
+    __android_log_assert("ort_api_base_fn != nullptr", "GenAI", "OrtGetApiBase not found");
+  }
+
+  const OrtApiBase* ort_api_base = ort_api_base_fn();
+  if (ort_api_base == nullptr) {
+    __android_log_assert("ort_api_base != nullptr", "GenAI", "OrtGetApiBase() returned nullptr");
+  }
+
+  // loop from the ORT version GenAI was built with, down to the minimum ORT version we require.
+  // as long as the libonnxruntime.so we loaded supports one of those we're good.
+  constexpr int genai_min_ort_api_version = 18;  // GenAI was first released around the time of ORT 1.18 so use that
+  for (int i = ORT_API_VERSION; i >= genai_min_ort_api_version; --i) {
+    api = ort_api_base->GetApi(i);
+    if (!api) {
+      __android_log_print(ANDROID_LOG_INFO, "GenAI", "ORT API Version %d was not found.", i);
+    } else {
+      __android_log_print(ANDROID_LOG_INFO, "GenAI", "ORT API Version %d was found", i);
+      break;
+    }
+  }
+
+  if (!api) {
+    __android_log_assert("api != nullptr", "GenAI",
+                         "%s did not have an ORT API version between %d and %d.",
+                         path.c_str(), ORT_API_VERSION, genai_min_ort_api_version);
+  }
+#else
   api = OrtGetApiBase()->GetApi(ORT_API_VERSION);
   if (!api)
     throw std::runtime_error("Onnxruntime is installed but is too old, please install a newer version");
+#endif
```

Comment:
nit: comments identifying the condition on `#else` and `#endif` are helpful when the original `#if` is farther away